pkgs/ok_http/lib/src/ok_http_web_socket.dart

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:jni/jni.dart';
+import 'package:web_socket/web_socket.dart';
+
+import 'jni/bindings.dart' as bindings;
+import 'ok_http_client.dart';
+
+/// A [WebSocket] implemented using the OkHttp library's
+/// [WebSocket](https://square.github.io/okhttp/5.x/okhttp/okhttp3/-web-socket/index.html)
+/// API.
+class OkHttpWebSocket implements WebSocket {
+  late bindings.OkHttpClient _client;
+  late final bindings.WebSocket _webSocket;
+  final _events = StreamController<WebSocketEvent>();
+  String? _protocol;
+
+  OkHttpWebSocket._() {
+    // Add the WSInterceptor to prevent response parsing errors.
+    _client = bindings.WSInterceptor.Companion
+        .addWSInterceptor(bindings.OkHttpClient_Builder())
+        .build();
+  }
+
+  /// Create a new WebSocket connection using `OkHttp`'s
+  /// [WebSocket](https://square.github.io/okhttp/5.x/okhttp/okhttp3/-web-socket/index.html)
+  /// API.
+  ///
+  /// The URL supplied in [url] must use the scheme ws or wss.
+  ///
+  /// If provided, the [protocols] argument indicates that subprotocols that
+  /// the peer is able to select. See
+  /// [RFC-6455 1.9](https://datatracker.ietf.org/doc/html/rfc6455#section-1.9).
+  static Future<WebSocket> connect(Uri url,
+          {Iterable<dynamic>? protocols}) async =>
+      OkHttpWebSocket._()._connect(url, protocols);
+
+  Future<WebSocket> _connect(Uri url, Iterable<dynamic>? protocols) async {
+    if (!url.isScheme('ws') && !url.isScheme('wss')) {
+      throw ArgumentError.value(
+          url, 'url', 'only ws: and wss: schemes are supported');
+    }
+
+    final requestBuilder =
+        bindings.Request_Builder().url1(url.toString().toJString());
+
+    if (protocols != null) {
+      requestBuilder.addHeader('Sec-WebSocket-Protocol'.toJString(),
+          protocols.join(', ').toJString());
+    }
+
+    var openCompleter = Completer<WebSocket>();
+
+    _client.newWebSocket(
+        requestBuilder.build(),
+        bindings.WebSocketListenerProxy(
+            bindings.WebSocketListenerProxy_WebSocketListener.implement(
+                bindings.$WebSocketListenerProxy_WebSocketListenerImpl(
+          onOpen: (webSocket, response) {
+            _webSocket = webSocket;
+
+            var protocolHeader =
+                response.header1('sec-websocket-protocol'.toJString());
+            if (!protocolHeader.isNull) {
+              _protocol = protocolHeader.toDartString();
+              if (!(protocols?.contains(_protocol) ?? true)) {
+                openCompleter
+                    .completeError(WebSocketException('Protocol mismatch. '
+                        'Expected one of $protocols, but received $_protocol'));
```

Comment:
Won't this get formatted weirdly if `protocols` is `null`?

Reply:
I guess not? If `protocols` is `null`, then the condition in the `if` block is `false` and hence the statements inside will not be executed.

Only when `protocols` is non-null and it does not contain the received `protocol` shall we throw the error.